### DISCUSSION TITLE
API Compatibility

### DISCUSSION THREAD

**Discussion by Zoruk** [dani-garcia/vaultwarden#4210] [Q&A]
Hi All,

I have noticed that some endpoint like `identity/accounts/prelogin` doesn't follow the exact same API as the official one in terms of case the project uses `PascalCase` and the official API uses `camelCase`. Is this voluntary ?

And otherwise since we can generate some code from the OpenAPI definitions is there any reason why the project doesn't use this ? 

Have a nice day all

**Comment by BlackDex** [comment]
Vaultwarden doesn't support the public API of Bitwarden, only a small subset of it to support the Directory Connector. So we are unable to generate any code from that. Also, there is no code generation for the Rocket Framework as far as i know, so that won't help either.

Regarding the `PascalCase` or `camelCase`, it actually doesn't matter what you send to Vaultwarden, we convert it internally to `PascalCase`. The reason for this is that in the past, the Bitwarden clients were sending a mixture of them. They should have converted all to `camelCase` as far as i know on what they send to the server. They do still allow receiving `PascalCase` for most API endpoints, and that is why we have not converted anything.

If any, we are not that strict in receiving either case formating, so that should be fine i think.

If you are only revering too what we output, then it's just that it is a big task to convert all to `camelCase`, but we probably need to at sometime in the future I think.

**Reply by Zoruk** [reply]
I totally agree here. 

The issue that I'm having it with auto-generated code the case sensitivity doesn't match and with some language like rust it just fails to parse the answers from `vaultwarden`.

Are you interested in some contribution to change the case in the responses ? If yes do you prefer to keep using the `json` macro or would you prefer to switch to `serde_json` ?

**Reply by BlackDex** [reply]
Well, the problem with that could be you would be breaking current implementations of others who use the `PascalCase` right now. We do use `serde_json` already.

I would suggest to maybe use the same technique we use to convert all to `PascalCase` to convert it to `camelCase`, it starts here https://github.com/dani-garcia/vaultwarden/blob/cbdcf8ef9f1ba0f4ad63f14d366ee778979a91ee/src/util.rs#L573

**Reply by Zoruk** [reply]
This would break implementation that are not compatible with the official API and from my understanding the aim of `vaultwarden` is to be compatible with the same API as the official one so that all tools works  with both implementations.

Here is a simple example with the differences
```
curl -XPOST -H "Content-Type: application/json" -d '{"email": "example@example.com"}' https://MY-VAULT-URI/identity/accounts/prelogin | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   103  100    71  100    32    217     98 --:--:-- --:--:-- --:--:--   316
{
  "Kdf": 0,
  "KdfIterations": 600000,
  "KdfMemory": null,
  "KdfParallelism": null
}

curl -XPOST -H "Content-Type: application/json" -d '{"email": "example@example.com"}' https://vault.bitwarden.eu/identity/accounts/prelogin | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   103    0    71  100    32    207     93 --:--:-- --:--:-- --:--:--   301
{
  "kdf": 0,
  "kdfIterations": 100000,
  "kdfMemory": null,
  "kdfParallelism": null
}
```

**Reply by dani-garcia** [reply]
The reason why we are using pascal case is because the server API was using it at some point, and eventually it got changed to the more common camel case. Note that the official clients do a fair amount of case conversion on each API response for compatibility:
https://github.com/bitwarden/clients/blob/f429756f68fad7ea0e696cfcc080240e510410b2/libs/common/src/models/response/base.response.ts#L8

The internal API is meant to be compatible with the official clients. Any other clients should follow the same case conversion strategy as the official clients for them to be considered compatible, imho.

We should eventually try to change the responses to match the official API, but it hasn't been a priority.

**Reply by BlackDex** [reply]
I understand the differences, that was clear, that is why i suggested the code part i linked to.

Breaking other implementations just because they are not following the API from Bitwarden but our, Vaultwarden implementation of the responses is not a valid argument i think.

Not that we shouldn't try to adjust it sometime, it will be something which could have a major impact and should be thoroughly tested on all clients from Bitwarden.

What I'm actually saying, a PR is good, but it's probably not something which we will merge rather quickly into a stable version.

**Reply by tessus** [reply]
Isn't https://github.com/dani-garcia/vaultwarden/issues/4196 related to this discussion?

**Reply by BlackDex** [reply]
Kinda, but for that i think it is the same. As long as the clients and server understand each other, i do not see the issue. Also, Bitwarden doesn't seem to be eager to modify on there side any time soon.

**Reply by Zoruk** [reply]
Also the `sdk` published on the `bitwarden` GitHub organization https://github.com/bitwarden/sdk/tree/main/crates/bitwarden-api-identity 

What I mean by this thread it that it would be nice to follow the official Open API extracted from the official server source. We can extract the OpenAPI definition using the method explained in the SDK Repository: https://github.com/bitwarden/sdk/tree/main#api-bindings

**Reply by BlackDex** [reply]
That is used for the Secrets Manager. Not the communication between clients and server. While it does have some endpoints in common like to login etc... It isn't specifically for the client-api

**Reply by BlackDex** [reply]
Though, there is a list of course of the other api's, I'm not sure what you want to do with that for Vaultwarden?

**Reply by Zoruk** [reply]
I wanted to play around with the API and maybe do a linux daemon which provide a bitwarden backend for libsecret. And trying to be compatible just for the authentication with both the official server and the vaultwarden implementation started to seams a bit too complicated to have two implementation supposed to follow the same open api reference.

**Reply by BlackDex** [reply]
It is very easy to convert it to `camelCase`, as @dani-garcia also mentioned, if  you really want to be compatible with the clients, you need to do this anyway. The conversion isn't that hard as i showed in the link.
I also created a lower-case-first version btw, maybe that helps, https://github.com/dani-garcia/vaultwarden/blob/cbdcf8ef9f1ba0f4ad63f14d366ee778979a91ee/src/util.rs#L708-L749

Else, see https://docs.rs/serde-aux/latest/serde_aux/container_attributes/fn.deserialize_struct_case_insensitive.html or https://serde.rs/attr-rename.html

**Comment by zacknewman** [comment]
> Isn't #4196 related to this discussion?

That's worse since that is a _WebAuthn_ violation. Bitwarden controls every part of _their_ API, but they don't control WebAuthn. Violating WebAuthn requires special care and should be avoided if possible. How many Vaultwarden devs have actually read the necessary standards (e.g., [WebAuthn Level 2 API](https://www.w3.org/TR/webauthn-2/)) and RFCs (e.g., [RFC 9053](https://www.rfc-editor.org/rfc/rfc9053))? Likely none or very few. The crate that is used even [makes several warnings](https://docs.rs/webauthn-rs-core/latest/webauthn_rs_core/core/index.html) of digging into the internals.

As the referenced issue shows, there is exactly one issue to be aware of by adhering to the spec and that is the web-vault violating WebAuthn by using `AttestationObject` and `clientDataJson` instead of `attestationObject` and `clientDataJSON` respectively as the spec requires, but that doesn't excuse Vaultwarden from also violating the spec.

The way that I recommend solving that problem is by patching the web-vault even though I am aware that Vaultwarden devs try to only patch the web-vault for things like CSS (as they should); however there have been [patches unrelated to just CSS](https://github.com/dani-garcia/bw_web_builds/blob/master/patches/v2023.12.0.patch). The reason to simply patch the web-vault such that those JSON keys are correctly formatted is the following:

* Only the web-vault can be used to register keys, and those JSON keys are only used during registration. This means patching this for the web-vault is the same thing as patching for all clients, so there is no worry that the patch only fixes the issue for one of the clients.
* The patch to the web-vault is stupid simple. Literally `sed -i 's/AttestationObject/attestationObject/g'` for the two files that matter and `sed -i 's/clientDataJson/clientDataJSON/g'` for the handful of files that need it.
* The amount of code currently used to deserialize `EnableWebauthnData` (including indirectly (e.g., the `Visitor` `util::UpCaseVisitor`)) is _substantially_ more than simply `derive`ing `Deserialize`.
    * This improves code audits which Vaultwarden is at a marked disadvantage compared to upstream.
    * `webauthn-rs` is the only dependency that is not updated frequently, and presumably one of the factors for this is the amount of code that has to be changed. By not relying on the internals as recommended by the project itself, updating the crate will be a _lot_ easier even though re-serialization of the data on the database is still a problem.
        * Once passkey in lieu of a master password is added to Bitwarden, `webauthn-rs` will have to be updated anyway so updating the crate is inevitable.

As the above points explain, that is why I think patching the web-vault makes a lot more sense especially since there is precedent (no matter how rare) to patching the web-vault for non-CSS so long as it makes sense; and simpler, more maintainable, spec-adhering, and easier-to-audit code in conjunction with a web-vault-only problem fits the bill for an exception to be made.

If patching the web-vault is still not done, then that doesn't excuse altering the entire payload but instead only the two problematic keys should be altered. This can still be achieved by not getting into the internals of `webauthn-rs` but does require more complex code. Specifically, a `Visitor` that reads the JSON payload and only alters `AttestationObject` and `clientDataJson` similar to what `UpCaseVisitor` is doing now.
